### PR TITLE
Cache result of MightHaveDispatchMap

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
Noticed this by accident.

We call this from multiple places within EEType construction, some of which are executed multiple times (especially the occurrence within `ComputeOptionalEETypeFields` that is called way more often than it really should be). We should just call once and cache the result. Computing it is expensive because it requires resolving all interface methods on the type to their implementation.

Saves about 2-3% of wallclock time within the AOT compiler.

### CODE REVIEW

**Review by jkotas — APPROVED** [review]
Thank you!

**Comment by MichalStrehovsky** [comment]
Sigh, the CI is all red because in optimized builds we sometimes create constructed EEType nodes for things that were not constructed before, just to check whether the node is marked - it isn't marked. I guess I'll need to make this `bool` a `bool?` and initialize lazily first time we need it.

I've already switched to other things, so this will wait here for a bit...

File: src/coreclr/src/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/CanonicalEETypeNode.cs [hunk 1]
```diff
@@ -44,7 +44,7 @@ protected override DependencyList ComputeNonRelocationBasedDependencies(NodeFact
 
             DefType closestDefType = _type.GetClosestDefType();
 
-            if (InterfaceDispatchMapNode.MightHaveInterfaceDispatchMap(_type, factory))
+            if (_mightHaveInterfaceDispatchMap)
```
Comment (MichalStrehovsky) [inline src/coreclr/src/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/CanonicalEETypeNode.cs:47]:
```suggestion
            if (MightHaveInterfaceDispatchMap(factory))
```

File: src/coreclr/src/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ConstructedEETypeNode.cs [hunk 2]
```diff
@@ -79,7 +79,7 @@ protected override DependencyList ComputeNonRelocationBasedDependencies(NodeFact
 
             DefType closestDefType = _type.GetClosestDefType();
 
-            if (InterfaceDispatchMapNode.MightHaveInterfaceDispatchMap(_type, factory))
+            if (_mightHaveInterfaceDispatchMap)
```
Comment (MichalStrehovsky) [inline src/coreclr/src/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ConstructedEETypeNode.cs:82]:
```suggestion
            if (MightHaveInterfaceDispatchMap(factory))
```

File: src/coreclr/src/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/EETypeNode.cs [hunk 3]
```diff
@@ -57,9 +57,10 @@ namespace ILCompiler.DependencyAnalysis
     /// </summary>
     public partial class EETypeNode : ObjectNode, IExportableSymbolNode, IEETypeNode, ISymbolDefinitionNode, ISymbolNodeWithLinkage
     {
-        protected TypeDesc _type;
-        internal EETypeOptionalFieldsBuilder _optionalFieldsBuilder = new EETypeOptionalFieldsBuilder();
-        internal EETypeOptionalFieldsNode _optionalFieldsNode;
+        protected readonly TypeDesc _type;
+        internal readonly EETypeOptionalFieldsBuilder _optionalFieldsBuilder = new EETypeOptionalFieldsBuilder();
+        internal readonly EETypeOptionalFieldsNode _optionalFieldsNode;
+        protected readonly bool _mightHaveInterfaceDispatchMap;
```
Comment (MichalStrehovsky) [inline src/coreclr/src/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/EETypeNode.cs:63]:
```suggestion
        protected bool? _mightHaveInterfaceDispatchMap;
```

File: src/coreclr/src/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/EETypeNode.cs [hunk 4]
```diff
@@ -72,6 +73,8 @@ public EETypeNode(NodeFactory factory, TypeDesc type)
             _type = type;
             _optionalFieldsNode = new EETypeOptionalFieldsNode(this);
 
+            _mightHaveInterfaceDispatchMap = EmitVirtualSlotsAndInterfaces && InterfaceDispatchMapNode.MightHaveInterfaceDispatchMap(type, factory);
+
```
Comment (MichalStrehovsky) [inline src/coreclr/src/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/EETypeNode.cs:77]:
```suggestion
```

File: src/coreclr/src/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/EETypeNode.cs [hunk 5]
```diff
@@ -882,7 +885,7 @@ private void OutputGenericInstantiationDetails(NodeFactory factory, ref ObjectDa
         /// </summary>
         protected internal virtual void ComputeOptionalEETypeFields(NodeFactory factory, bool relocsOnly)
         {
-            if (!relocsOnly && EmitVirtualSlotsAndInterfaces && InterfaceDispatchMapNode.MightHaveInterfaceDispatchMap(_type, factory))
+            if (!relocsOnly && _mightHaveInterfaceDispatchMap)
```
Comment (MichalStrehovsky) [inline src/coreclr/src/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/EETypeNode.cs:888]:
```suggestion
            if (!relocsOnly && MightHaveInterfaceDispatchMap(factory))
```

File: src/coreclr/src/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/EETypeNode.cs [hunk 6]
```diff
@@ -72,6 +73,8 @@ public EETypeNode(NodeFactory factory, TypeDesc type)
             _type = type;
             _optionalFieldsNode = new EETypeOptionalFieldsNode(this);
 
+            _mightHaveInterfaceDispatchMap = EmitVirtualSlotsAndInterfaces && InterfaceDispatchMapNode.MightHaveInterfaceDispatchMap(type, factory);
+
             factory.TypeSystemContext.EnsureLoadableType(type);
         }
```
Comment (MichalStrehovsky) [inline src/coreclr/src/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/EETypeNode.cs:79]:
```suggestion
        }
        
        protected bool MightHaveInterfaceDispatchMap(NodeFactory factory)
        {
            if (!_mightHaveInterfaceDispatchMap.HasValue)
            {
                _mightHaveInterfaceDispatchMap = EmitVirtualSlotsAndInterfaces && InterfaceDispatchMapNode.MightHaveInterfaceDispatchMap(_type, factory);
            }
            
            return _mightHaveInterfaceDispatchMap.Value;
        }
```

**Pull request merged**